devices/src/virtio/vsock/unix/muxer_killq.rs

```diff
@@ -0,0 +1,137 @@
+/// to check wether any connections need to be terminated.
+/// When the kill queue becomes full, though, pushing fails, so connections that should be terminated
+/// are left out. The queue is not synchronized anymore. When that happens, the muxer will first drain
+/// the queue, and then replace it with a new queue, created by walking the connection pool,
+/// looking for connections that will be expiring in the future.
+///
+use std::collections::{HashMap, VecDeque};
+use std::time::Instant;
+
+use super::defs;
+use super::muxer::ConnMapKey;
+use super::MuxerConnection;
+
+/// A kill queue item, holding the connection key and the scheduled time for termination.
+///
+#[derive(Clone, Copy)]
+struct MuxerKillQItem {
+    key: ConnMapKey,
+    kill_time: Instant,
+}
+
+/// The connection kill queue: a FIFO structure, storing the connections that are scheduled for
+/// termination.
+///
+pub struct MuxerKillQ {
+    /// The kill queue contents.
+    q: VecDeque<MuxerKillQItem>,
+
+    /// The kill queue sync status:
+    /// - when true, all connections that are awaiting termination are guaranteed to be in this queue;
+    /// - when false, some connections may have been left out.
+    ///
+    synced: bool,
+}
+
+impl MuxerKillQ {
+    const SIZE: usize = defs::MUXER_KILLQ_SIZE;
+
+    /// Trivial kill queue constructor.
+    ///
+    pub fn new() -> Self {
+        Self {
+            q: VecDeque::with_capacity(Self::SIZE),
+            synced: true,
+        }
+    }
+
+    /// Create a kill queue by walking the connection pool, looking for connections that are set to
+    /// expire at some point in the future.
+    /// Note: if more than `Self::SIZE` connections are found, the queue will be created in an
+    ///       out-of-sync state, and will be discarded after it is emptied.
+    ///
+    pub fn from_conn_map(conn_map: &HashMap<ConnMapKey, MuxerConnection>) -> Self {
+        let mut q_buf: Vec<MuxerKillQItem> = Vec::with_capacity(Self::SIZE);
+        let mut synced = true;
+        for (key, conn) in conn_map.iter() {
+            if !conn.will_expire() {
+                continue;
+            }
+            if q_buf.len() >= Self::SIZE {
+                synced = false;
+                break;
+            }
+            q_buf.push(MuxerKillQItem {
+                key: *key,
+                kill_time: conn.expiry().unwrap(),
+            });
+        }
+        q_buf.sort_unstable_by_key(|it| it.kill_time);
+        Self {
+            q: q_buf.into(),
+            synced,
+        }
+    }
+
+    /// Push a connection key to the queue, scheduling it for termination at `CONN_SHUTDOWN_TIMEOUT_MS`
+    /// from now (the push time).
+    ///
+    pub fn push(&mut self, key: ConnMapKey, kill_time: Instant) {
+        if !self.is_synced() || self.is_full() {
+            self.synced = false;
+            return;
+        }
+        self.q.push_back(MuxerKillQItem { key, kill_time });
```

Comment:
Yes, connections are added to the kill queue as they get scheduled for the axe. Expiration time being a constant, that would maintain the queue order. However, even if the order does get mangled somehow, that's not critical, since we're under no contract to provide strict termination time.